### PR TITLE
Tuples & iterators

### PR DESCRIPTION
**Pull request by dvc94ch**

### CODE REVIEW

File: src/js.rs [hunk 1]
```diff
@@ -645,15 +683,21 @@ impl WasmMultiValueShim {
...
+                        match field.ty.num() {
+                            NumType::F32 => ret.push_str("f32 "),
+                            NumType::F64 => ret.push_str("f64 "),
+                            _ => {
+                                if size > 4 {
+                                    ret.push_str("i64 ");
+                                } else {
+                                    ret.push_str("i32 ");
+                                }
+                            }
```
Comment (wngr) [inline src/js.rs:700]:
```suggestion
                        let r = match field.ty.num() {
                            NumType::F32 => "f32 ",
                            NumType::F64 => "f64 ",
                            _ if size > 4 => "i64 ",
                            _ => "i32 ",
                            };
                         ret.push_str(r);
```

File: tests/functions.rs [hunk 2]
```diff
@@ -541,7 +513,93 @@ compile_pass! {
...
+    export class Api {
+        constructor();
+
+        fetch(url, imports): Promise<void>;
+
+        tuple0(): void;
+
+        tuple1(arg0: number): number;
+
+        tuple2(arg0: number, arg1: number): [number,number,];
```
Comment (wngr) [inline tests/functions.rs:569]: ew. will look into that.

**Review by wngr — COMMENTED** [review]
looking into js iterators.

seems `result_future::js` is also failing

File: src/rust.rs [hunk 3]
```diff
@@ -464,6 +542,22 @@ impl RustGenerator {
             NumType::F64 => quote!(f64),
         }
     }
+
+    fn ffi_num_type(&self, ty: NumType) -> rust::Tokens {
```
Comment (dvc94ch) [inline src/rust.rs:546]: here we handle it

File: src/rust.rs [hunk 4]
```diff
@@ -234,14 +252,23 @@ impl RustGenerator {
         // the first argument.
         quote! {
             #[no_mangle]
-            pub extern "C" fn #name(_: #(self.num_type(self.abi.iptr())), boxed: #(self.num_type(self.abi.iptr()))) {
+            pub extern "C" fn #name(_: #(self.ffi_num_type(self.abi.iptr())), boxed: #(self.ffi_num_type(self.abi.iptr()))) {
```
Comment (dvc94ch) [inline src/rust.rs:255]: and here

File: src/js.rs [hunk 5]
```diff
@@ -772,12 +829,13 @@ pub mod test_runner {
...
-        let (_, p) = runner_file.keep()?;
+        //        js_file.keep()?;
+        //        rust_file.keep()?;
+        //        library_file.keep()?;
+        //        let (_, p) = runner_file.keep()?;
         let test = TestCases::new();
-        test.pass(p);
+        //        test.pass(p);
+        test.pass(runner_file.as_ref());
         Ok(())
```
Comment (dvc94ch) [inline src/js.rs:839]: what's going on here?

Comment (wngr) [reply src/js.rs:839]: I accidentally committed the `keep` stuff from when I was debugging. Undid that, but I find it convenient so I kept it around as comments.

**Review by wngr — APPROVED** [review]

**Comment by bitfriend** [comment]
Can you please provide the example about tuple as function return in `api.rsh`?

**Pull request merged**